zebra-scanner/src/tests.rs

```diff
@@ -0,0 +1,313 @@
+//! Test that we can scan blocks.
+//!
+
+use std::sync::Arc;
+
+use zcash_client_backend::{
+    proto::compact_formats::{
+        self as compact, ChainMetadata, CompactBlock, CompactSaplingOutput, CompactSaplingSpend,
+        CompactTx,
+    },
+    scanning::scan_block,
+};
+use zcash_note_encryption::Domain;
+use zcash_primitives::{
+    block::BlockHash,
+    consensus::{BlockHeight, Network},
+    constants::SPENDING_KEY_GENERATOR,
+    memo::MemoBytes,
+    sapling::{
+        note_encryption::{sapling_note_encryption, SaplingDomain},
+        util::generate_random_rseed,
+        value::NoteValue,
+        Note, Nullifier, SaplingIvk,
+    },
+    zip32::{AccountId, DiversifiableFullViewingKey, ExtendedSpendingKey},
+};
+
+use color_eyre::Result;
+
+use rand::{rngs::OsRng, RngCore};
+
+use ff::{Field, PrimeField};
+use group::GroupEncoding;
+use zebra_chain::{
+    block::{Block, Height},
+    serialization::ZcashSerialize,
+    transaction::Transaction,
+};
+
+#[tokio::test]
+#[allow(clippy::print_stdout)]
+async fn scanning_from_zebra() -> Result<()> {
+    let account = AccountId::from(12);
+    let extsk = ExtendedSpendingKey::master(&[]);
+    let dfvk: DiversifiableFullViewingKey = extsk.to_diversifiable_full_viewing_key();
+    let vks: Vec<(&AccountId, &SaplingIvk)> = vec![];
+    let nf = Nullifier([7; 32]);
+
+    let (state_config, network) = Default::default();
+    let (_state_service, read_only_state_service, _latest_chain_tip, _chain_tip_change) =
+        zebra_state::spawn_init(state_config, network, Height::MAX, 3000).await?;
+    let db = read_only_state_service.db();
+
+    let mut height = Height(0);
```

Comment:
This will fail with:

```
---- tests::scanning_from_zebra stdout ----
thread 'tests::scanning_from_zebra' panicked at zebra-scanner/src/tests.rs:85:10:
called `Result::unwrap()` on an `Err` value: TreeSizeUnknown { protocol: Sapling, at_height: BlockHeight(0) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So it seems that we need to start scanning in sapling:

```suggestion
    let mut height = zebra_chain::parameters::NetworkUpgrade::Sapling.activation_height(network).unwrap();
```

But, the state is empty in this test so we need a synchronized chain to make this loop run or some alternative.

Reply:
Ok, so i fixed this at https://github.com/ZcashFoundation/zebra/pull/7758/commits/cee1db69864ff1b820893b868008d0b0ff0a4155 with a fake sapling size of 1. 

Also, i changed to use a populated cache where we have 11 blocks inside, this allow us to test these 11 blocks are converted to compactblock and scanned. No relevant transactions should be found.

I also separated the test in 2 tests where one do the above and another use fake generated blocks to get a successful case.